### PR TITLE
Mining: Prevent slowdown in CreateNewBlock on large mempools

### PR DESCRIPTION
**Pull request by sdaftuar**
I've been investigating performance regressions in CreateNewBlock.

`addPackageTxs()` is supposed to "fail fast" for transactions that don't fit in a block.  However while we skip the most expensive ancestor/descendant calculations for failing transactions (generally), it turns out the other work we're doing (mostly map lookups?) can still be pretty slow.

After trying various approaches to speed up those operations  (replacing maps with unordered_maps, changing the sort order on the ancestor score function to avoid needless re-sorting, and even getting rid of the maps altogether in favor of storing set information directly in the mempool entry), the one optimization that dominates all these is to just return earlier when the block is nearly full.

So in this PR: when we're within 4000 weight of the block being full, if we consider and fail to add 1000 transactions in a row, then give up.  I've benchmarked this as reducing the average run of CNB from ~~84ms to 63ms~~ 74ms to 61ms, with negligible difference in fees.  Most of CNB time is currently taken by the call to TestBlockValidity, which is unaffected by this PR; so focusing just on package selection, the improvement from this change is a reduction in average `addPackageTxs()` time from 19ms to 7ms, over the time period I analyzed (first week of December, 2016).

I also added some commits that provide benchmarking of CNB when running with -debug=bench, which I thought might be generally useful/interesting.

### CODE REVIEW

**Comment by sdaftuar** [comment]
Update OP with a correction and a more relevant statistic.

File: src/miner.cpp [hunk 1]
```diff
@@ -177,7 +179,11 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
     // transaction (which in most cases can be a no-op).
     fIncludeWitness = IsWitnessEnabled(pindexPrev, chainparams.GetConsensus());
 
-    addPackageTxs();
+    int nPackagesSelected = 0;
+    int nDescendantsUpdated = 0;
```
Comment (luke-jr) [inline src/miner.cpp]: Might make more sense to put these on the class, rather than pass by-reference?

File: src/miner.cpp [hunk 2]
```diff
@@ -358,6 +370,12 @@ void BlockAssembler::addPackageTxs()
 
     CTxMemPool::indexed_transaction_set::index<ancestor_score>::type::iterator mi = mempool.mapTx.get<ancestor_score>().begin();
     CTxMemPool::txiter iter;
+
+    // Limit the number of attempts to add transactions to the block when it is
+    // close to full.
+    const int64_t MAX_CONSECUTIVE_FAILURES = 1000;
```
Comment (luke-jr) [inline src/miner.cpp]: The old code used 50 for this. Does 1000 work good enough in practice?

File: src/miner.cpp [hunk 3]
```diff
@@ -419,6 +437,14 @@ void BlockAssembler::addPackageTxs()
                 mapModifiedTx.get<ancestor_score>().erase(modit);
                 failedTx.insert(iter);
             }
+
+            ++nConsecutiveFailed;
+
+            if (nConsecutiveFailed > MAX_CONSECUTIVE_FAILURES && nBlockWeight >
+                    nBlockMaxWeight - 4000) {
```
Comment (luke-jr) [inline src/miner.cpp]: Perhaps the additional condition(s) should be before incrementing, or we could end up just aborting immediately as we approach the max block weight even if we could easily fit a few more txs in.

**Review by luke-jr — APPROVED** [review]
utACK, certainly an improvement either way. (this behaviour used to exist before the CNB refactoring, maybe the old code is worth looking at)

**Comment by TheBlueMatt** [comment]
utACK 03a407662b008fe030d81bb54dfd3b677b248392, didnt bother to benchmark, should be obvious wins, even if minor.

File: src/miner.cpp (hunk 1 above)
Comment (sdaftuar) [inline src/miner.cpp]: I'm working on a refactor of the `BlockAssembler` members separately (as part of a patch to exclude recent transactions from new blocks, when the fee difference is negligible), so I'd prefer to defer the decision of whether to include this in the class until I'm ready to PR that patch.

File: src/miner.cpp (hunk 2 above)
Comment (sdaftuar) [inline src/miner.cpp]: Yes, well I also tried setting this to 100 and the performance was not really distinguishable.

File: src/miner.cpp (hunk 3 above)
Comment (sdaftuar) [inline src/miner.cpp]: I'm not following; every time we add a new tx to the block we reset the counter...  Can you clarify?

**Comment by TheBlueMatt** [comment]
Is this worth backporting?

**Comment by sdaftuar** [comment]
@thebluematt I don't feel strongly about whether we backport this change, to be honest (though it's simple enough that I don't think there's any downside).  

But it would be nice to merge this into master if there are no further comments, so I can continue work on additional mining changes (to support excluding recently received transactions if the fee difference from doing so is below some threshold).

File: src/miner.cpp [hunk 4]
```diff
@@ -439,6 +465,9 @@ void BlockAssembler::addPackageTxs()
             continue;
         }
 
+        // This transaction will make it in; reset the failed counter.
+        nConsecutiveFailed = 0;
```
Comment (JeremyRubin) [inline src/miner.cpp]: Could you add more of a comment on why resetting is correct behavior here?

It seems to me at first glance, that if we fail to add something earlier, we should continue to tally those as failures?
Perhaps:
```
Assuming that below a certain threshold `T` of probability
(i.e.,`T = P(success at M | failure at M-1....0)` where `M = 1000`) 
of adding something to the block we want to give up,
 we expect that 
`1000>D>0. P(success at N+D+1 | failure at N+D,... success at N, failure at N-1, ...) 
> P(success at N+D | failure at N, failure at N-1, ...)`?
```

Maybe not the most accurate, but would be helpful for future reviewers trying to understand what the intention is.

**Review by JeremyRubin — APPROVED** [review]
utack. I've reviewed the changes and it seems reasonable. @sdaftuar I also left some ideas for a few complexity optimizations you may not have yet considered for the expensive loop; but those are for future work :).

File: src/miner.cpp [hunk 5]
```diff
@@ -282,16 +291,18 @@ void BlockAssembler::AddToBlock(CTxMemPool::txiter iter)
...
-void BlockAssembler::UpdatePackagesForAdded(const CTxMemPool::setEntries& alreadyAdded,
+int BlockAssembler::UpdatePackagesForAdded(const CTxMemPool::setEntries& alreadyAdded,
         indexed_modified_transaction_set &mapModifiedTx)
 {
+    int nDescendants = 0;
     BOOST_FOREACH(const CTxMemPool::txiter it, alreadyAdded) {
         CTxMemPool::setEntries descendants;
         mempool.CalculateDescendants(it, descendants);
         // Insert all descendants (not yet in block) into the modified set
         BOOST_FOREACH(CTxMemPool::txiter desc, descendants) {
```
Comment (sdaftuar) [inline src/miner.cpp]: Thanks, will try some of these out in future optimization efforts.

File: src/miner.cpp (hunk 4 above)
Comment (sdaftuar) [inline src/miner.cpp]: This is incremented whenever we fail (not just for failures after we're above a certain block weight), so it needs to be reset when adding a new tx.

**Comment by sdaftuar** [comment]
Addressed @JeremyRubin's nits.

**Comment by JeremyRubin** [comment]
re-utack 4d1eb10

**Comment by sdaftuar** [comment]
Squashed 4d1eb10 -> 011124a

**Comment by TheBlueMatt** [comment]
re-utACK 011124a2b278c5a60bad5f1b0b4abbf7ebc95aa0

**Review by gmaxwell — APPROVED** [review]
utACK

**Comment by laanwj** [comment]
Removing 'needs backport' label as a backport exists (#10127)

**Pull request merged**